content/blog/automation-api-as-platform/index.md

```diff
@@ -0,0 +1,275 @@
+
+```typescript
+const projectName = "pulumi_over_http";
+
+const createPulumiProgram = (content: string) => async () => {
+    // Create a bucket and expose a website index document
+    const siteBucket = new s3.Bucket("s3-website-bucket", {
+        website: {
+            indexDocument: "index.html",
+        },
+    });
+```
+
+This part of the function lets users write content.
+
+```typescript
+    // Our HTML is defined based on what the caller sends.
+    const indexContent = content;
+
+    // write our index.html into the site bucket
+    let object = new s3.BucketObject("index", {
+        bucket: siteBucket,
+        content: indexContent,
+        contentType: "text/html; charset=utf-8",
+        key: "index.html"
+    });
+```
+
+To make the contents accessible, we create and set a policy for the S3 bucket.
+
+```typescript
+    // Create an S3 Bucket Policy to allow public read of all objects in bucket
+    function publicReadPolicyForBucket(bucketName): PolicyDocument {
+        return {
+            Version: "2012-10-17",
+            Statement: [{
+                Effect: "Allow",
+                Principal: "*",
+                Action: [
+                    "s3:GetObject"
+                ],
+                Resource: [
+                    `arn:aws:s3:::${bucketName}/*` // policy refers to bucket name explicitly
+                ]
+            }]
+        };
+    }
+
+    // Set the access policy for the bucket so all objects are readable
+    let bucketPolicy = new s3.BucketPolicy("bucketPolicy", {
+        bucket: siteBucket.bucket, // refer to the bucket created earlier
+        policy: siteBucket.bucket.apply(publicReadPolicyForBucket) // use output property `siteBucket.bucket`
+    });
+
+    return {
+        websiteUrl: siteBucket.websiteEndpoint,
+    };
+};
+```
+
+### Building the REST API
+
+We use the Express `RequestHandler` callback to create the S3 static website buckets by creating the stack locally. Calling this handler creates a new bucket which represents an individual site.
+
+```typescript
+// creates new sites
+const createHandler: express.RequestHandler = async (req, res) => {
+    const stackName = req.body.id;
+    const content = req.body.content as string;
+    try {
+        // create a new stack
+        const stack = await LocalWorkspace.createStack({
+            stackName,
+            projectName,
+            // generate our pulumi program on the fly from the POST body
+            program: createPulumiProgram(content),
+        });
+        await stack.setConfig("aws:region", { value: "us-west-2" });
+        // deploy the stack, tailing the logs to console
+        const upRes = await stack.up({ onOutput: console.info });
+        res.json({ id: stackName, url: upRes.outputs.websiteUrl.value });
+    } catch (e) {
+        if (e instanceof StackAlreadyExistsError) {
```

Comment:
I'm not sure if it makes sense to discuss given the scope and focus of the article, but the ability to detect and handle structured errors is pretty important. Automation API gives you tools to detect and handle error scenarios which is a necessary part of "putting your infrastructure online" or embedding it in other complex software.

Reply:
good point to highlight